plenum/server/view_change/instance_change_provider.py

```diff
@@ -0,0 +1,118 @@
+        self._time_provider = time_provider
+        self._instance_change_db = instance_change_db
+        self._fill_cache_by_db()
+
+    def add_vote(self, msg: InstanceChange, voter: str):
+        view_no = msg.viewNo
+        vote = Vote(timestamp=self._time_provider(),
+                    reason=msg.reason)
+        # add to cache
+        self._cache.add(view_no, voter, vote)
+        # add to db
+        self._update_db_from_cache(view_no)
+
+    def has_view(self, view_no: int) -> bool:
+        self._update_votes(view_no)
+        return view_no in self._cache
+
+    def has_inst_chng_from(self, view_no: int, voter: str) -> bool:
+        self._update_votes(view_no)
+        return view_no in self._cache and voter in self._cache[view_no]
+
+    def has_quorum(self, view_no: int, quorum: int) -> bool:
+        self._update_votes(view_no)
+        return view_no in self._cache and len(self._cache[view_no]) >= quorum
+
+    def remove_view(self, view_to_remove: int):
+        for view_no in sorted(self._cache.keys()):
+            if view_no > view_to_remove:
+                break
+            del self._cache[view_no]
+            if self._instance_change_db:
+                self._instance_change_db.remove(str(view_no))
+
+    def items(self):
+        return self._cache.items()
+
+    def _update_votes(self, view_no: int):
+        if self._outdated_ic_interval <= 0 or view_no not in self._cache:
+            return
+        db_need_update = False
+        for voter, vote in dict(self._cache[view_no]).items():
+            now = self._time_provider()
+            if vote.timestamp < now - self._outdated_ic_interval:
+                logger.info("InstanceChangeProvider: Discard InstanceChange from {} for ViewNo {} "
+                            "because it is out of date (was received {}sec "
+                            "ago)".format(voter, view_no, int(now - vote.timestamp)))
+                self._cache.remove_vote(view_no, voter)
+                db_need_update = True
+        if db_need_update:
+            self._update_db_from_cache(view_no)
+
+    def _update_db_from_cache(self, view_no):
+        if not self._instance_change_db:
+            return
+        serialized_value = \
+            instance_change_db_serializer.serialize(self._cache.get(view_no, None))
+        if serialized_value:
+            self._instance_change_db.put(str(view_no), serialized_value)
+
+    def _fill_cache_by_db(self):
+        if not self._instance_change_db:
+            return
+        for view_no, serialized_votes in self._instance_change_db.iterator(include_value=True):
+            if not view_no.isdigit():
+                logger.warning("InstanceChangeProvider: view_no='{}' "
+                               "must be of int type".format(view_no))
+                break
+            votes_as_dict = instance_change_db_serializer.deserialize(serialized_votes)
+            if not votes_as_dict:
+                break
```

Comment:
What if we had serialized empty votes for some `view_no`? This will prevent from processing all other votes stored in DB for other view_nos. Did you mean `continue` here instead of `break`?

Reply:
Added break for case when ```serialized_votes`` is ``None```. I think If we found an incorrect entry, processing for other entries shouldn't be stopped.

Reply:
But correctly serialized empty dict will be also correctly deserialized as an empty dict, triggering this condition.

Reply:
Yes. And I think this is correct behavior because we shouldn't store in _cache view_no without votes. Maybe will be better to remove this view_no from the DB, but I think it's not a good practice because on the start a node can be works incorrect and removing something from DB which can lead to problem in the next node starting.